File: rpc/python/src/caravan_rpc/resources/queue.py
"""MessageQueue seam + Redis/RabbitMQ/SQS impls.

The ``MessageQueue`` seam is ``@wagon``-decorated so callers dispatch through
``client(MessageQueue).publish(...)``. Caravan picks the backend per yaml
composition (oss-local redis-streams or rabbitmq; cloud-managed SQS).

The ``topic:`` parameter is a portability device — Redis Streams + RabbitMQ
use it as a queue/stream name; SQS ignores it (each SqsQueue instance is
bound to a single queue URL because that's how SQS works).
"""

from __future__ import annotations

import json
import os
import time
import uuid
from urllib.parse import urlparse

from .. import wagon


@wagon
class MessageQueue:
    """Caravan-owned seam for at-least-once message queues."""

    def publish(self, topic: str, message: dict) -> str:
        """Publish a message. Returns the broker-assigned message ID."""
        ...

    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        """Pull-style consume. Returns up to ``count`` (id, payload) pairs,
        blocking up to ``block_ms`` for messages to arrive."""
        ...

    def ack(self, topic: str, message_id: str) -> None:
        """Acknowledge / delete a previously consumed message."""
        ...

    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None:
        """Extend the message's invisibility window. No-op on backends that
        don't support it (Redis Streams, RabbitMQ)."""
        ...


class RedisStreamQueue:
    """Redis Streams with consumer groups. Caravan emits this for oss-local
    queue/redis-streams (default) and queue/redis variants."""

    def __init__(self, url: str, consumer_group: str = "caravan-workers") -> None:
        try:
            import redis
        except ImportError as e:
            raise ImportError(
                "RedisStreamQueue requires redis. Install `caravan-rpc[redis]`."
            ) from e
        self._client = redis.Redis.from_url(url, decode_responses=True)
        self._consumer_group = consumer_group
        self._consumer_name = f"worker-{uuid.uuid4().hex[:8]}"
        self._initialized_groups: set[str] = set()

    def _ensure_group(self, topic: str) -> None:
        if topic in self._initialized_groups:
            return
        import redis as _redis

        try:
            self._client.xgroup_create(topic, self._consumer_group, id="0", mkstream=True)
        except _redis.ResponseError as e:
            if "BUSYGROUP" not in str(e):
                raise
        self._initialized_groups.add(topic)

    def publish(self, topic: str, message: dict) -> str:
        self._ensure_group(topic)
        msg_id: str = self._client.xadd(topic, {"data": json.dumps(message)})
        return msg_id

    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        self._ensure_group(topic)
        results = self._client.xreadgroup(
            groupname=self._consumer_group,
            consumername=self._consumer_name,
            streams={topic: ">"},
            count=count,
            block=block_ms,
        )
        if not results:
            return []
        out: list[tuple[str, dict]] = []
        for _stream, entries in results:
            for msg_id, fields in entries:
                out.append((msg_id, json.loads(fields["data"])))
        return out

    def ack(self, topic: str, message_id: str) -> None:
        self._client.xack(topic, self._consumer_group, message_id)

    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None:
        # Redis Streams have no visibility timeout — pending entries stay
        # claimed by the consumer until XACK or manual XCLAIM by reaper.
        return None


class RabbitMQQueue:
    """RabbitMQ via pika (synchronous channel). Caravan emits this for
    oss-local queue/rabbitmq composition."""

    def __init__(self, url: str) -> None:
        try:
            import pika  # noqa: F401 — checked at construction; actual import lazy in _channel_
        except ImportError as e:
            raise ImportError(
                "RabbitMQQueue requires pika. Install `caravan-rpc[rabbit]`."
            ) from e
        scheme = urlparse(url).scheme
        if scheme not in ("amqp", "amqps"):
            raise ValueError(
                f"RabbitMQQueue expects amqp://... or amqps://...; got scheme {scheme!r}"
            )
        self._url = url
        self._connection = None
        self._channel = None
        self._declared_queues: set[str] = set()

    @classmethod
    def from_url(cls, url: str) -> RabbitMQQueue:
        return cls(url)

    def _channel_(self):
        if self._channel is not None and not self._channel.is_closed:
            return self._channel
        import pika

        params = pika.URLParameters(self._url)
        self._connection = pika.BlockingConnection(params)
        self._channel = self._connection.channel()
        return self._channel

    def _ensure_queue(self, topic: str) -> None:
        if topic in self._declared_queues:
            return
        ch = self._channel_()
        ch.queue_declare(queue=topic, durable=True)
        self._declared_queues.add(topic)

    def publish(self, topic: str, message: dict) -> str:
        import pika

        self._ensure_queue(topic)
        ch = self._channel_()
        msg_id = uuid.uuid4().hex
        ch.basic_publish(
            exchange="",
            routing_key=topic,
            body=json.dumps(message).encode(),
            properties=pika.BasicProperties(
                message_id=msg_id, delivery_mode=2, content_type="application/json"
            ),
        )
        return msg_id

    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        self._ensure_queue(topic)
        ch = self._channel_()
        deadline = time.monotonic() + block_ms / 1000.0
        out: list[tuple[str, dict]] = []
        while len(out) < count:
            method, _props, body = ch.basic_get(queue=topic, auto_ack=False)
            if method is None:
                if time.monotonic() >= deadline:
                    break
                time.sleep(0.05)
                continue
            out.append((str(method.delivery_tag), json.loads(body.decode())))
        return out

    def ack(self, topic: str, message_id: str) -> None:
        ch = self._channel_()
        ch.basic_ack(delivery_tag=int(message_id))

    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None:
        # RabbitMQ has no SQS-style visibility timeout; unacked messages
        # requeue when the channel closes.
        return None


class SqsQueue:
    """AWS SQS via boto3. Caravan emits this for cloud-managed queue
    composition (M4-cloud hybrid-dev and onward).

    SQS is single-queue-per-URL — the ``topic:`` parameter on publish/consume
    is ignored. Each ``aws_sqs_queue`` Caravan emits gets its own SqsQueue
    instance via ``from_url``.
    """

    def __init__(self, queue_url: str) -> None:
        try:
            import boto3
        except ImportError as e:
            raise ImportError(
                "SqsQueue requires boto3. Install `caravan-rpc[aws]`."
            ) from e
        self._queue_url = queue_url
        self._client = boto3.client("sqs")

    @classmethod
    def from_url(cls, url: str) -> SqsQueue:
        scheme = urlparse(url).scheme
        if scheme not in ("http", "https"):
            raise ValueError(
                f"SqsQueue expects an https://sqs.<region>.amazonaws.com/... URL; got scheme {scheme!r}"
            )
        return cls(url)

    @classmethod
    def from_env(cls) -> SqsQueue:
        url = os.environ.get("QUEUE_URL")
        if not url:
            raise ValueError("SqsQueue.from_env requires QUEUE_URL to be set")
        return cls.from_url(url)

    def publish(self, topic: str, message: dict) -> str:
        # topic ignored — single-queue substrate.
        resp = self._client.send_message(
            QueueUrl=self._queue_url, MessageBody=json.dumps(message)
        )
        return resp["MessageId"]

    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        wait_time = max(0, min(20, block_ms // 1000))  # SQS long-poll 0-20s.
        resp = self._client.receive_message(
            QueueUrl=self._queue_url,
            MaxNumberOfMessages=min(count, 10),  # SQS hard cap.
            WaitTimeSeconds=wait_time,
        )
        out: list[tuple[str, dict]] = []
        for m in resp.get("Messages", []):
            out.append((m["ReceiptHandle"], json.loads(m["Body"])))
        return out

    def ack(self, topic: str, message_id: str) -> None:
        # message_id is the ReceiptHandle from consume.
        self._client.delete_message(QueueUrl=self._queue_url, ReceiptHandle=message_id)

    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None:
        self._client.change_message_visibility(
            QueueUrl=self._queue_url,
            ReceiptHandle=message_id,
            VisibilityTimeout=seconds,
        )

File: rpc/rust/caravan-rpc/src/resources/queue.rs
//! MessageQueue seam + Redis / RabbitMQ / SQS impls.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

use crate::wagon;

/// Errors from the MessageQueue seam. Serde-serializable for over-the-
/// wire dispatch; `serde_json::Error` is converted to `Json(String)` at
/// construction time.
#[derive(Debug, Error, Serialize, Deserialize, Clone)]
pub enum QueueError {
    #[error("JSON error: {0}")]
    Json(String),
    #[error("Redis error: {0}")]
    Redis(String),
    #[error("RabbitMQ error: {0}")]
    RabbitMQ(String),
    #[error("SQS error: {0}")]
    Sqs(String),
    #[error("invalid URL: {0}")]
    InvalidUrl(String),
}

impl From<serde_json::Error> for QueueError {
    fn from(e: serde_json::Error) -> Self {
        QueueError::Json(e.to_string())
    }
}

/// Caravan-owned seam for at-least-once message queues. Decorated with
/// `#[wagon]` so callers dispatch via `client::<dyn MessageQueue>()`.
///
/// The `topic:` parameter is a portability device: Redis Streams +
/// RabbitMQ use it as a queue/stream name; SQS ignores it (each SQS
/// queue URL is its own substrate).
#[wagon]
pub trait MessageQueue: Send + Sync {
    fn publish(&self, topic: &str, message: &Value) -> Result<String, QueueError>;
    fn consume(
        &self,
        topic: &str,
        count: usize,
        block_ms: usize,
    ) -> Result<Vec<(String, Value)>, QueueError>;
    fn ack(&self, topic: &str, message_id: &str) -> Result<(), QueueError>;
    fn extend_visibility(
        &self,
        topic: &str,
        message_id: &str,
        seconds: u64,
    ) -> Result<(), QueueError>;
}

// --- Redis Streams impl (gated behind `resources-redis`) -------------------

#[cfg(feature = "resources-redis")]
pub use redis_impl::RedisStreamQueue;

#[cfg(feature = "resources-redis")]
mod redis_impl {
    use redis::streams::{StreamReadOptions, StreamReadReply};
    use redis::{Commands, RedisError};
    use serde_json::Value;

    use super::{MessageQueue, QueueError};

    impl From<RedisError> for QueueError {
        fn from(e: RedisError) -> Self {
            QueueError::Redis(e.to_string())
        }
    }

    pub struct RedisStreamQueue {
        client: redis::Client,
        consumer_group: String,
        consumer_name: String,
    }

    impl RedisStreamQueue {
        pub fn new(url: &str, consumer_group: &str) -> Result<Self, QueueError> {
            let client = redis::Client::open(url).map_err(QueueError::from)?;
            let consumer_name = format!("worker-{}", &uuid::Uuid::new_v4().to_string()[..8]);
            Ok(Self {
                client,
                consumer_group: consumer_group.to_string(),
                consumer_name,
            })
        }

        fn ensure_group(
            &self,
            conn: &mut redis::Connection,
            topic: &str,
        ) -> Result<(), QueueError> {
            let result: Result<(), RedisError> = redis::cmd("XGROUP")
                .arg("CREATE")
                .arg(topic)
                .arg(&self.consumer_group)
                .arg("0")
                .arg("MKSTREAM")
                .query(conn);
            match result {
                Ok(()) => Ok(()),
                Err(e) if e.to_string().contains("BUSYGROUP") => Ok(()),
                Err(e) => Err(QueueError::Redis(e.to_string())),
            }
        }
    }

    impl MessageQueue for RedisStreamQueue {
        fn publish(&self, topic: &str, message: &Value) -> Result<String, QueueError> {
            let mut conn = self.client.get_connection()?;
            self.ensure_group(&mut conn, topic)?;
            let payload = serde_json::to_string(message)?;
            let id: String = conn.xadd(topic, "*", &[("data", &payload)])?;
            Ok(id)
        }

        fn consume(
            &self,
            topic: &str,
            count: usize,
            block_ms: usize,
        ) -> Result<Vec<(String, Value)>, QueueError> {
            let mut conn = self.client.get_connection()?;
            self.ensure_group(&mut conn, topic)?;
            let opts = StreamReadOptions::default()
                .group(&self.consumer_group, &self.consumer_name)
                .count(count)
                .block(block_ms);
            let reply: StreamReadReply = conn.xread_options(&[topic], &[">"], &opts)?;
            let mut messages = Vec::new();
            for key in reply.keys {
                for entry in key.ids {
                    if let Some(redis::Value::BulkString(bytes)) = entry.map.get("data") {
                        let data_str = String::from_utf8_lossy(bytes);
                        let value: Value = serde_json::from_str(&data_str)?;
                        messages.push((entry.id, value));
                    }
                }
            }
            Ok(messages)
        }

        fn ack(&self, topic: &str, message_id: &str) -> Result<(), QueueError> {
            let mut conn = self.client.get_connection()?;
            let _: () = conn.xack(topic, &self.consumer_group, &[message_id])?;
            Ok(())
        }

        fn extend_visibility(
            &self,
            _topic: &str,
            _message_id: &str,
            _seconds: u64,
        ) -> Result<(), QueueError> {
            Ok(()) // No SQS-style visibility timeout in Redis Streams.
        }
    }
}

// --- RabbitMQ impl (gated behind `resources-rabbit`) -----------------------

#[cfg(feature = "resources-rabbit")]
pub use rabbit_impl::RabbitMQQueue;

#[cfg(feature = "resources-rabbit")]
mod rabbit_impl {
    use std::sync::Mutex;

    use serde_json::Value;

    use super::{MessageQueue, QueueError};

    pub struct RabbitMQQueue {
        url: String,
        state: Mutex<RabbitMQState>,
    }

    struct RabbitMQState {
        connection: Option<lapin::Connection>,
        channel: Option<lapin::Channel>,
        declared: std::collections::HashSet<String>,
    }

    impl RabbitMQQueue {
        pub fn new(url: &str) -> Result<Self, QueueError> {
            let scheme = url::Url::parse(url)
                .map_err(|e| QueueError::InvalidUrl(e.to_string()))?
                .scheme()
                .to_string();
            if scheme != "amqp" && scheme != "amqps" {
                return Err(QueueError::InvalidUrl(format!(
                    "RabbitMQQueue expects amqp(s)://; got {scheme}"
                )));
            }
            Ok(Self {
                url: url.to_string(),
                state: Mutex::new(RabbitMQState {
                    connection: None,
                    channel: None,
                    declared: std::collections::HashSet::new(),
                }),
            })
        }

        fn ensure_channel(&self) -> Result<lapin::Channel, QueueError> {
            let mut state = self.state.lock().unwrap();
            if let Some(ch) = &state.channel {
                if ch.status().connected() {
                    return Ok(ch.clone());
                }
            }
            let url = self.url.clone();
            let (conn, ch) = tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    let conn =
                        lapin::Connection::connect(&url, lapin::ConnectionProperties::default())
                            .await
                            .map_err(|e| QueueError::RabbitMQ(e.to_string()))?;
                    let ch = conn
                        .create_channel()
                        .await
                        .map_err(|e| QueueError::RabbitMQ(e.to_string()))?;
                    Ok::<_, QueueError>((conn, ch))
                })
            })?;
            state.connection = Some(conn);
            state.channel = Some(ch.clone());
            state.declared.clear();
            Ok(ch)
        }

        fn ensure_queue(&self, ch: &lapin::Channel, topic: &str) -> Result<(), QueueError> {
            {
                let state = self.state.lock().unwrap();
                if state.declared.contains(topic) {
                    return Ok(());
                }
            }
            let topic_owned = topic.to_string();
            let ch_clone = ch.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    ch_clone
                        .queue_declare(
                            &topic_owned,
                            lapin::options::QueueDeclareOptions {
                                durable: true,
                                ..Default::default()
                            },
                            lapin::types::FieldTable::default(),
                        )
                        .await
                        .map_err(|e| QueueError::RabbitMQ(e.to_string()))
                        .map(|_| ())
                })
            })?;
            let mut state = self.state.lock().unwrap();
            state.declared.insert(topic.to_string());
            Ok(())
        }
    }

    impl MessageQueue for RabbitMQQueue {
        fn publish(&self, topic: &str, message: &Value) -> Result<String, QueueError> {
            let ch = self.ensure_channel()?;
            self.ensure_queue(&ch, topic)?;
            let body = serde_json::to_vec(message)?;
            let msg_id = uuid::Uuid::new_v4().to_string().replace('-', "");
            let topic_owned = topic.to_string();
            let msg_id_owned = msg_id.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    ch.basic_publish(
                        "",
                        &topic_owned,
                        lapin::options::BasicPublishOptions::default(),
                        &body,
                        lapin::BasicProperties::default()
                            .with_message_id(msg_id_owned.into())
                            .with_delivery_mode(2)
                            .with_content_type("application/json".into()),
                    )
                    .await
                    .map_err(|e| QueueError::RabbitMQ(e.to_string()))?
                    .await
                    .map_err(|e| QueueError::RabbitMQ(e.to_string()))?;
                    Ok::<_, QueueError>(())
                })
            })?;
            Ok(msg_id)
        }

        fn consume(
            &self,
            topic: &str,
            count: usize,
            block_ms: usize,
        ) -> Result<Vec<(String, Value)>, QueueError> {
            let ch = self.ensure_channel()?;
            self.ensure_queue(&ch, topic)?;
            let deadline =
                std::time::Instant::now() + std::time::Duration::from_millis(block_ms as u64);
            let topic_owned = topic.to_string();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    let mut out: Vec<(String, Value)> = Vec::new();
                    while out.len() < count {
                        if std::time::Instant::now() >= deadline {
                            break;
                        }
                        let msg = ch
                            .basic_get(&topic_owned, lapin::options::BasicGetOptions::default())
                            .await
                            .map_err(|e| QueueError::RabbitMQ(e.to_string()))?;
                        match msg {
                            Some(delivery) => {
                                let payload: Value = serde_json::from_slice(&delivery.data)?;
                                out.push((delivery.delivery_tag.to_string(), payload));
                            }
                            None => {
                                tokio::time::sleep(std::time::Duration::from_millis(50)).await;
                            }
                        }
                    }
                    Ok::<_, QueueError>(out)
                })
            })
        }

        fn ack(&self, _topic: &str, message_id: &str) -> Result<(), QueueError> {
            let ch = self.ensure_channel()?;
            let delivery_tag: u64 = message_id
                .parse()
                .map_err(|e: std::num::ParseIntError| QueueError::RabbitMQ(e.to_string()))?;
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    ch.basic_ack(delivery_tag, lapin::options::BasicAckOptions::default())
                        .await
                        .map_err(|e| QueueError::RabbitMQ(e.to_string()))
                })
            })
        }

        fn extend_visibility(
            &self,
            _topic: &str,
            _message_id: &str,
            _seconds: u64,
        ) -> Result<(), QueueError> {
            Ok(())
        }
    }
}

// --- SQS impl (gated behind `resources-aws`) -------------------------------

#[cfg(feature = "resources-aws")]
pub use sqs_impl::SqsQueue;

#[cfg(feature = "resources-aws")]
mod sqs_impl {
    use serde_json::Value;

    use super::{MessageQueue, QueueError};

    /// AWS SQS impl via aws-sdk-sqs. Single-queue per URL.
    pub struct SqsQueue {
        client: aws_sdk_sqs::Client,
        queue_url: String,
    }

    impl SqsQueue {
        pub fn from_url(url: &str) -> Result<Self, QueueError> {
            let scheme = url::Url::parse(url)
                .map_err(|e| QueueError::InvalidUrl(e.to_string()))?
                .scheme()
                .to_string();
            if scheme != "http" && scheme != "https" {
                return Err(QueueError::InvalidUrl(format!(
                    "SqsQueue expects https://sqs.<region>.amazonaws.com/...; got {scheme}"
                )));
            }
            let client = tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    let cfg = aws_config::defaults(aws_config::BehaviorVersion::latest())
                        .load()
                        .await;
                    aws_sdk_sqs::Client::new(&cfg)
                })
            });
            Ok(Self {
                client,
                queue_url: url.to_string(),
            })
        }

        pub fn from_env() -> Result<Self, QueueError> {
            let url = std::env::var("QUEUE_URL")
                .map_err(|_| QueueError::Sqs("SqsQueue::from_env requires QUEUE_URL".into()))?;
            Self::from_url(&url)
        }
    }

    impl MessageQueue for SqsQueue {
        fn publish(&self, _topic: &str, message: &Value) -> Result<String, QueueError> {
            let body = serde_json::to_string(message)?;
            let queue_url = self.queue_url.clone();
            let client = self.client.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    let resp = client
                        .send_message()
                        .queue_url(&queue_url)
                        .message_body(body)
                        .send()
                        .await
                        .map_err(|e| QueueError::Sqs(e.to_string()))?;
                    Ok(resp.message_id().unwrap_or_default().to_string())
                })
            })
        }

        fn consume(
            &self,
            _topic: &str,
            count: usize,
            block_ms: usize,
        ) -> Result<Vec<(String, Value)>, QueueError> {
            let wait_seconds = (block_ms / 1000).clamp(0, 20) as i32;
            let max_msgs = count.min(10) as i32;
            let queue_url = self.queue_url.clone();
            let client = self.client.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    let resp = client
                        .receive_message()
                        .queue_url(&queue_url)
                        .max_number_of_messages(max_msgs)
                        .wait_time_seconds(wait_seconds)
                        .send()
                        .await
                        .map_err(|e| QueueError::Sqs(e.to_string()))?;
                    let mut out: Vec<(String, Value)> = Vec::new();
                    for m in resp.messages.unwrap_or_default() {
                        let handle = m
                            .receipt_handle
                            .ok_or_else(|| QueueError::Sqs("missing ReceiptHandle".into()))?;
                        let body = m
                            .body
                            .ok_or_else(|| QueueError::Sqs("missing Body".into()))?;
                        let value: Value = serde_json::from_str(&body)?;
                        out.push((handle, value));
                    }
                    Ok(out)
                })
            })
        }

        fn ack(&self, _topic: &str, message_id: &str) -> Result<(), QueueError> {
            let queue_url = self.queue_url.clone();
            let handle = message_id.to_string();
            let client = self.client.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    client
                        .delete_message()
                        .queue_url(&queue_url)
                        .receipt_handle(handle)
                        .send()
                        .await
                        .map_err(|e| QueueError::Sqs(e.to_string()))
                        .map(|_| ())
                })
            })
        }

        fn extend_visibility(
            &self,
            _topic: &str,
            message_id: &str,
            seconds: u64,
        ) -> Result<(), QueueError> {
            let queue_url = self.queue_url.clone();
            let handle = message_id.to_string();
            let client = self.client.clone();
            tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    client
                        .change_message_visibility()
                        .queue_url(&queue_url)
                        .receipt_handle(handle)
                        .visibility_timeout(seconds as i32)
                        .send()
                        .await
                        .map_err(|e| QueueError::Sqs(e.to_string()))
                        .map(|_| ())
                })
            })
        }
    }
}
